File: ios/Kwift/com/lightningkite/kwift/actuals/Date.swift
//
//  Date.swift
//  KwiftTemplate
//
//  Created by Joseph Ivie on 9/10/19.
//  Copyright © 2019 Joseph Ivie. All rights reserved.
//

import Foundation


public extension Date {
    var time: Int64 {
        return Int64((self.timeIntervalSince1970 * 1000.0).rounded())
    }
    
    func iso8601() -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
        return formatter.string(from: self)
    }

    init(_ milliseconds:Int64) {
        self = Date(timeIntervalSince1970: TimeInterval(milliseconds) / 1000)
    }

    func before(_ other: Date) -> Bool {
        return self.time < other.time
    }
    func after(_ other: Date) -> Bool {
        return self.time > other.time
    }
    var dayOfWeek: Int32 { return Int32(Calendar.current.component(.weekday, from: self)) }
    var dayOfMonth: Int32 { return Int32(Calendar.current.component(.day, from: self)) }
    var monthOfYear: Int32 { return Int32(Calendar.current.component(.month, from: self)) }
    var yearAd: Int32 { return Int32(Calendar.current.component(.year, from: self)) }
    var hourOfDay: Int32 { return Int32(Calendar.current.component(.hour, from: self)) }
    var minuteOfHour: Int32 { return Int32(Calendar.current.component(.minute, from: self)) }
    var secondOfMinute: Int32 { return Int32(Calendar.current.component(.second, from: self)) }
    var dateAlone: DateAlone {
        return DateAlone(yearAd, monthOfYear, dayOfMonth)
    }
    var timeAlone: TimeAlone {
        return TimeAlone(hourOfDay, minuteOfHour, secondOfMinute)
    }

    func sameDay(_ other: Date) -> Bool {
        return self.yearAd == other.yearAd && self.monthOfYear == other.monthOfYear && self.dayOfMonth == other.dayOfMonth
    }
    func sameMonth(_ other: Date) -> Bool {
        return self.yearAd == other.yearAd && self.monthOfYear == other.monthOfYear
    }
    func sameYear(_ other: Date) -> Bool {
        return self.yearAd == other.yearAd
    }

    func dayOfWeek(_ value: Int32, _ existing: Date? = nil) -> Date {
        let components = Calendar.current.dateComponents([.weekday], from: self)
        let diff = Int(value) - components.weekday!
        return self + diff.days()
    }
    func dayOfMonth(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.year, .month, .hour, .minute, .second, .nanosecond], from: self)
        components.day = Int(value)
        return Calendar.current.date(from: components)!
    }
    func monthOfYear(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.year, .day, .hour, .minute, .second, .nanosecond], from: self)
        components.month = Int(value)
        return Calendar.current.date(from: components)!
    }
    func yearAd(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.month, .day, .hour, .minute, .second, .nanosecond], from: self)
        components.year = Int(value)
        return Calendar.current.date(from: components)!
    }
    func hourOfDay(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.year, .month, .day, .minute, .second, .nanosecond], from: self)
        components.hour = Int(value)
        return Calendar.current.date(from: components)!
    }
    func minuteOfHour(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.year, .month, .day, .hour, .second, .nanosecond], from: self)
        components.minute = Int(value)
        return Calendar.current.date(from: components)!
    }
    func secondOfMinute(_ value: Int32, _ existing: Date? = nil) -> Date {
        var components = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute, .nanosecond], from: self)
        components.second = Int(value)
        return Calendar.current.date(from: components)!
    }
    func addDayOfWeek(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .weekday, value: Int(value), to: self)!
    }
    func addDayOfMonth(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .day, value: Int(value), to: self)!
    }
    func addMonthOfYear(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .month, value: Int(value), to: self)!
    }
    func addYearAd(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .year, value: Int(value), to: self)!
    }
    func addHourOfDay(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .hour, value: Int(value), to: self)!
    }
    func addMinuteOfHour(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .minute, value: Int(value), to: self)!
    }
    func addSecondOfMinute(_ value: Int32, _ existing: Date? = nil) -> Date {
        return Calendar.current.date(byAdding: .second, value: Int(value), to: self)!
    }

    mutating func setDayOfWeek(_ value: Int32) -> Date { self = self.dayOfWeek(value); return self }
    mutating func setDayOfMonth(_ value: Int32) -> Date { self = self.dayOfMonth(value); return self }
    mutating func setMonthOfYear(_ value: Int32) -> Date { self = self.monthOfYear(value); return self }
    mutating func setYearAd(_ value: Int32) -> Date { self = self.yearAd(value); return self }
    mutating func setHourOfDay(_ value: Int32) -> Date { self = self.hourOfDay(value); return self }
    mutating func setMinuteOfHour(_ value: Int32) -> Date { self = self.minuteOfHour(value); return self }
    mutating func setSecondOfMinute(_ value: Int32) -> Date { self = self.secondOfMinute(value); return self }
    mutating func setAddDayOfWeek(_ value: Int32) -> Date { self = self.addDayOfWeek(value); return self }
    mutating func setAddDayOfMonth(_ value: Int32) -> Date { self = self.addDayOfMonth(value); return self }
    mutating func setAddMonthOfYear(_ value: Int32) -> Date { self = self.addMonthOfYear(value); return self }
    mutating func setAddYearAd(_ value: Int32) -> Date { self = self.addYearAd(value); return self }
    mutating func setAddHourOfDay(_ value: Int32) -> Date { self = self.addHourOfDay(value); return self }
    mutating func setAddMinuteOfHour(_ value: Int32) -> Date { self = self.addMinuteOfHour(value); return self }
    mutating func setAddSecondOfMinute(_ value: Int32) -> Date { self = self.addSecondOfMinute(value); return self }
    
    mutating func set(_ dateAlone: DateAlone) -> Date {
        var components = Calendar.current.dateComponents([.hour, .minute, .second, .nanosecond], from: self)
        components.year = Int(dateAlone.year)
        components.month = Int(dateAlone.month)
        components.day = Int(dateAlone.day)
        self = Calendar.current.date(from: components)!
        return self
    }
    mutating func set(_ timeAlone: TimeAlone) -> Date {
        var components = Calendar.current.dateComponents([.year, .month, .day], from: self)
        components.hour = Int(timeAlone.hour)
        components.minute = Int(timeAlone.minute)
        components.second = Int(timeAlone.second)
        self = Calendar.current.date(from: components)!
        return self
    }
    mutating func set(_ dateAlone: DateAlone, _ timeAlone: TimeAlone) -> Date {
        var components = Calendar.current.dateComponents([], from: self)
        components.year = Int(dateAlone.year)
        components.month = Int(dateAlone.month)
        components.day = Int(dateAlone.day)
        components.hour = Int(timeAlone.hour)
        components.minute = Int(timeAlone.minute)
        components.second = Int(timeAlone.second)
        self = Calendar.current.date(from: components)!
        return self
    }
    
    func format(_ dateStyle: ClockPartSize, _ timeStyle: ClockPartSize) -> String {
        return format(dateStyle: dateStyle, timeStyle: timeStyle)
    }
    func format(dateStyle: ClockPartSize, timeStyle: ClockPartSize) -> String {
        var rawDateStyle: DateFormatter.Style = .none
        switch dateStyle {
            case .None: rawDateStyle = .none
            case .Short: rawDateStyle = .short
            case .Medium: rawDateStyle = .medium
            case .Long: rawDateStyle = .long
            case .Full: rawDateStyle = .full
        }
        var rawTimeStyle: DateFormatter.Style = .none
        switch timeStyle {
            case .None: rawTimeStyle = .none
            case .Short: rawTimeStyle = .short
            case .Medium: rawTimeStyle = .medium
            case .Long: rawTimeStyle = .long
            case .Full: rawTimeStyle = .full
        }
        
        let formatter = DateFormatter()
        formatter.dateStyle = rawDateStyle
        formatter.timeStyle = rawTimeStyle
        return formatter.string(from: self)
    }
}

public extension DateAlone {
    func set(_ date: Date) -> DateAlone {
        let components = Calendar.current.dateComponents([.year, .month, .day], from: date)
        self.year = Int32(components.year!)
        self.month = Int32(components.month!)
        self.day = Int32(components.day!)
        return self
    }
    func setDayOfWeek(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).dayOfWeek(value)) }
    func setDayOfMonth(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).dayOfMonth(value)) }
    func setMonthOfYear(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).monthOfYear(value)) }
    func setYearAd(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).yearAd(value)) }
    func setAddDayOfWeek(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).addDayOfWeek(value)) }
    func setAddDayOfMonth(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).addDayOfMonth(value)) }
    func setAddMonthOfYear(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).addMonthOfYear(value)) }
    func setAddYearAd(_ value: Int32) -> DateAlone { return set(dateFrom(self, TimeAlone.noon).addYearAd(value)) }
    
    func dayOfWeek(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).dayOfWeek(value).dateAlone }
    func dayOfMonth(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).dayOfMonth(value).dateAlone }
    func monthOfYear(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).monthOfYear(value).dateAlone }
    func yearAd(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).yearAd(value).dateAlone }
    func addDayOfWeek(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).addDayOfWeek(value).dateAlone }
    func addDayOfMonth(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).addDayOfMonth(value).dateAlone }
    func addMonthOfYear(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).addMonthOfYear(value).dateAlone }
    func addYearAd(_ value: Int32) -> DateAlone { return dateFrom(self, TimeAlone.noon).addYearAd(value).dateAlone }
}

public extension TimeAlone {
    func set(_ date: Date) -> TimeAlone {
        let components = Calendar.current.dateComponents([.hour, .minute, .second], from: date)
        self.hour = Int32(components.hour!)
        self.minute = Int32(components.minute!)
        self.second = Int32(components.second!)
        return self
    }
}

public func dateFrom(_ dateAlone: DateAlone, _ timeAlone: TimeAlone) -> Date {
    return dateFrom(dateAlone: dateAlone, timeAlone: timeAlone)
}
public func dateFrom(dateAlone: DateAlone, timeAlone: TimeAlone) -> Date {
    let components = DateComponents(
        year: Int(dateAlone.year),
        month: Int(dateAlone.month),
        day: Int(dateAlone.day),
        hour: Int(timeAlone.hour),
        minute: Int(timeAlone.minute),
        second: Int(timeAlone.second),
        nanosecond: 0
    )
    return Calendar.current.date(from: components)!
}

public extension Int32 {
    func milliseconds() -> TimeInterval { return TimeInterval(self) / 1000 }
    func seconds() -> TimeInterval { return TimeInterval(self) }
    func minutes() -> TimeInterval { return TimeInterval(self * 60) }
    func hours() -> TimeInterval { return TimeInterval(self * 60 * 60) }
    func days() -> TimeInterval { return TimeInterval(self * 60 * 60 * 24) }
}

public extension Int {
    func milliseconds() -> TimeInterval { return TimeInterval(self) / 1000 }
    func seconds() -> TimeInterval { return TimeInterval(self) }
    func minutes() -> TimeInterval { return TimeInterval(self * 60) }
    func hours() -> TimeInterval { return TimeInterval(self * 60 * 60) }
    func days() -> TimeInterval { return TimeInterval(self * 60 * 60 * 24) }
}

public extension TimeInterval {
    var milliseconds: Int64 {
        return Int64(self * 1000)
    }
}

public enum TimeNames {
    private static var formatter = DateFormatter()
    public static var monthNames = formatter.monthSymbols!
    public static var shortMonthNames = formatter.shortMonthSymbols!
    public static var weekdayNames = formatter.weekdaySymbols!
    public static var shortWeekdayNames = formatter.shortWeekdaySymbols!
    
    public static func monthName(_ oneIndexedPosition: Int32) -> String { return monthNames[oneIndexedPosition - 1] }
    public static func shortMonthName(_ oneIndexedPosition: Int32) -> String { return shortMonthNames[oneIndexedPosition - 1] }
    public static func weekdayName(_ oneIndexedPosition: Int32) -> String { return weekdayNames[oneIndexedPosition - 1] }
    public static func shortWeekdayName(_ oneIndexedPosition: Int32) -> String { return shortWeekdayNames[oneIndexedPosition - 1] }
}

public extension TimeAlone {
    static func - (left: TimeAlone, right: TimeAlone) -> TimeAlone{
        let result = left.hour * 60 * 60 + left.minute * 60 + left.second + right.hour * 60 * 60 + right.minute * 60 + right.second
        if result < 0{
            return TimeAlone(0,0,0)
        }else{
            return TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
        }
    }

    static func + (left: TimeAlone, right: TimeAlone) -> TimeAlone{
        let result = (left.hour * 60 * 60 + left.minute * 60 + left.second) + (right.hour * 60 * 60 + right.minute * 60 + right.second)
        return TimeAlone(result / 60 / 60,  result / 60 % 60, result % 60)
    }
    
    static func - (left: TimeAlone, right: TimeInterval) -> TimeAlone{
        let result = left.hour * 60 * 60 + left.minute * 60 + left.second - Int32(right)
        if result < 0{
            return TimeAlone(0,0,0)
        }else{
            return TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
        }
    }

    static func + (left: TimeAlone, right: TimeInterval) -> TimeAlone{
        let result = (left.hour * 60 * 60 + left.minute * 60 + left.second) + Int32(right)
        return TimeAlone(result / 60 / 60,  result / 60 % 60, result % 60)
    }

}

File: android/src/main/java/com/lightningkite/kwift/actual/Date.kt
package com.lightningkite.kwift.actual

import com.lightningkite.kwift.shared.ClockPartSize
import com.lightningkite.kwift.shared.DateAlone
import com.lightningkite.kwift.shared.TimeAlone
import java.text.DateFormat
import java.text.DateFormatSymbols
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

@Deprecated("Use from shared package instead", ReplaceWith("TimeAlone", "com.lightningkite.kwift.shared.TimeAlone"))
typealias TimeAlone = com.lightningkite.kwift.shared.TimeAlone

@Deprecated("Use from shared package instead", ReplaceWith("DateAlone", "com.lightningkite.kwift.shared.DateAlone"))
typealias DateAlone = com.lightningkite.kwift.shared.DateAlone

@Deprecated(
    "Use from shared package instead",
    ReplaceWith("ClockPartSize", "com.lightningkite.kwift.shared.ClockPartSize")
)
typealias ClockPartSize = com.lightningkite.kwift.shared.ClockPartSize

inline class TimeInterval(val milliseconds: Long) {
    val seconds: Double get() = milliseconds / 1000.0
}

private val calendarPool = ConcurrentLinkedQueue<Calendar>()
private inline fun <T> useCalendar(action: (Calendar) -> T): T {
    val cal = calendarPool.poll() ?: Calendar.getInstance()
    val result = action(cal)
    calendarPool.offer(cal)
    return result
}

private val datePool = ConcurrentLinkedQueue<Date>()
private inline fun <T> useDate(action: (Date) -> T): T {
    val cal = datePool.poll() ?: Date()
    val result = action(cal)
    datePool.offer(cal)
    return result
}

fun Int.milliseconds(): TimeInterval = TimeInterval(this.toLong())
fun Int.seconds(): TimeInterval = TimeInterval(this.toLong() * 1000L)
fun Int.minutes(): TimeInterval = TimeInterval(this.toLong() * 60L * 1000L)
fun Int.hours(): TimeInterval = TimeInterval(this.toLong() * 60L * 60L * 1000L)
fun Int.days(): TimeInterval = TimeInterval(this.toLong() * 24L * 60L * 60L * 1000L)

operator fun Date.plus(interval: TimeInterval): Date = Date(time + interval.milliseconds)
operator fun Date.minus(interval: TimeInterval): Date = Date(time - interval.milliseconds)

/**One Indexed**/
val Date.dayOfWeek: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.DAY_OF_WEEK) }
/**One Indexed**/
val Date.dayOfMonth: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.DAY_OF_MONTH) }
/**One Indexed**/
val Date.monthOfYear: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.MONTH) + 1 }
val Date.yearAd: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.YEAR) }
val Date.hourOfDay: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.HOUR_OF_DAY) }
val Date.minuteOfHour: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.MINUTE) }
val Date.secondOfMinute: Int get() = useCalendar { it.timeInMillis = time; it.get(Calendar.SECOND) }

val Date.dateAlone: DateAlone
    get() {
        return useCalendar { cal ->
            cal.timeInMillis = time
            DateAlone(
                cal.get(Calendar.YEAR),
                cal.get(Calendar.MONTH) + 1,
                cal.get(Calendar.DAY_OF_MONTH)
            )
        }
    }
val Date.timeAlone: TimeAlone
    get() {
        return useCalendar { cal ->
            cal.timeInMillis = time
            TimeAlone(
                cal.get(Calendar.HOUR_OF_DAY),
                cal.get(Calendar.MINUTE),
                cal.get(Calendar.SECOND)
            )
        }
    }

fun Date.sameDay(other: Date): Boolean {
    return this.yearAd == other.yearAd && this.monthOfYear == other.monthOfYear && this.dayOfMonth == other.dayOfMonth
}

fun Date.sameMonth(other: Date): Boolean {
    return this.yearAd == other.yearAd && this.monthOfYear == other.monthOfYear
}

fun Date.sameYear(other: Date): Boolean {
    return this.yearAd == other.yearAd
}

/**One Indexed**/
fun Date.dayOfWeek(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.DAY_OF_WEEK, value)
        existing.time = it.timeInMillis
        existing
    }

/**One Indexed**/
fun Date.dayOfMonth(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.DAY_OF_MONTH, value)
        existing.time = it.timeInMillis
        existing
    }

/**One Indexed**/
fun Date.monthOfYear(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.MONTH, value - 1)
        existing.time = it.timeInMillis
        existing
    }

fun Date.yearAd(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.YEAR, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.hourOfDay(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.HOUR_OF_DAY, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.minuteOfHour(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.MINUTE, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.secondOfMinute(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.SECOND, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addDayOfWeek(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.DAY_OF_WEEK, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addDayOfMonth(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.DAY_OF_MONTH, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addMonthOfYear(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.MONTH, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addYearAd(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.YEAR, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addHourOfDay(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.HOUR_OF_DAY, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addMinuteOfHour(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.MINUTE, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.addSecondOfMinute(value: Int, existing: Date = Date()) =
    useCalendar {
        it.timeInMillis = time
        it.add(Calendar.SECOND, value)
        existing.time = it.timeInMillis
        existing
    }

fun Date.setDayOfWeek(value: Int) = this.dayOfWeek(value, existing = this)
fun Date.setDayOfMonth(value: Int) = this.dayOfMonth(value, existing = this)
fun Date.setMonthOfYear(value: Int) = this.monthOfYear(value, existing = this)
fun Date.setYearAd(value: Int) = this.yearAd(value, existing = this)
fun Date.setHourOfDay(value: Int) = this.hourOfDay(value, existing = this)
fun Date.setMinuteOfHour(value: Int) = this.minuteOfHour(value, existing = this)
fun Date.setSecondOfMinute(value: Int) = this.secondOfMinute(value, existing = this)
fun Date.setAddDayOfWeek(value: Int) = this.addDayOfWeek(value, existing = this)
fun Date.setAddDayOfMonth(value: Int) = this.addDayOfMonth(value, existing = this)
fun Date.setAddMonthOfYear(value: Int) = this.addMonthOfYear(value, existing = this)
fun Date.setAddYearAd(value: Int) = this.addYearAd(value, existing = this)
fun Date.setAddHourOfDay(value: Int) = this.addHourOfDay(value, existing = this)
fun Date.setAddMinuteOfHour(value: Int) = this.addMinuteOfHour(value, existing = this)
fun Date.setAddSecondOfMinute(value: Int) = this.addSecondOfMinute(value, existing = this)

fun Date.set(dateAlone: DateAlone): Date {
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.YEAR, dateAlone.year)
        it.set(Calendar.MONTH, dateAlone.month - 1)
        it.set(Calendar.DAY_OF_MONTH, dateAlone.day)
        this.time = it.timeInMillis
    }
    return this
}

fun Date.set(timeAlone: TimeAlone): Date {
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.HOUR_OF_DAY, timeAlone.hour)
        it.set(Calendar.MINUTE, timeAlone.minute)
        it.set(Calendar.SECOND, timeAlone.second)
        this.time = it.timeInMillis
    }
    return this
}

fun Date.set(dateAlone: DateAlone, timeAlone: TimeAlone): Date {
    useCalendar {
        it.timeInMillis = time
        it.set(Calendar.YEAR, dateAlone.year)
        it.set(Calendar.MONTH, dateAlone.month - 1)
        it.set(Calendar.DAY_OF_MONTH, dateAlone.day)
        it.set(Calendar.HOUR_OF_DAY, timeAlone.hour)
        it.set(Calendar.MINUTE, timeAlone.minute)
        it.set(Calendar.SECOND, timeAlone.second)
        this.time = it.timeInMillis
    }
    return this
}

fun DateAlone.set(date: Date): DateAlone {
    useCalendar {
        it.timeInMillis = date.time
        year = it.get(Calendar.YEAR)
        month = it.get(Calendar.MONTH) + 1
        day = it.get(Calendar.DAY_OF_MONTH)
    }
    return this
}

fun TimeAlone.set(date: Date): TimeAlone {
    useCalendar {
        it.timeInMillis = date.time
        hour = it.get(Calendar.HOUR_OF_DAY)
        minute = it.get(Calendar.MINUTE)
        second = it.get(Calendar.SECOND)
    }
    return this
}

fun DateAlone.setDayOfMonth(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setDayOfMonth(value); set(date) } }

fun DateAlone.setMonthOfYear(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setMonthOfYear(value); set(date) } }

fun DateAlone.setYearAd(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setYearAd(value); set(date) } }

fun DateAlone.setDayOfWeek(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setDayOfWeek(value); set(date) } }

fun DateAlone.setAddDayOfWeek(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setAddDayOfWeek(value); set(date) } }

fun DateAlone.setAddDayOfMonth(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setAddDayOfMonth(value); set(date) } }

fun DateAlone.setAddMonthOfYear(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setAddMonthOfYear(value); set(date) } }

fun DateAlone.setAddYearAd(value: Int): DateAlone =
    apply { useDate { date -> date.set(this); date.setAddYearAd(value); set(date) } }


fun DateAlone.dayOfMonth(value: Int): DateAlone = useDate { date -> date.set(this); date.setDayOfMonth(value); date.dateAlone }
fun DateAlone.monthOfYear(value: Int): DateAlone = useDate { date -> date.set(this); date.setMonthOfYear(value); date.dateAlone }
fun DateAlone.yearAd(value: Int): DateAlone = useDate { date -> date.set(this); date.setYearAd(value); date.dateAlone }
fun DateAlone.dayOfWeek(value: Int): DateAlone = useDate { date -> date.set(this); date.setDayOfWeek(value); date.dateAlone }
fun DateAlone.addDayOfWeek(value: Int): DateAlone = useDate { date -> date.set(this); date.setAddDayOfWeek(value); date.dateAlone }
fun DateAlone.addDayOfMonth(value: Int): DateAlone = useDate { date -> date.set(this); date.setAddDayOfMonth(value); date.dateAlone }
fun DateAlone.addMonthOfYear(value: Int): DateAlone = useDate { date -> date.set(this); date.setAddMonthOfYear(value); date.dateAlone }
fun DateAlone.addYearAd(value: Int): DateAlone = useDate { date -> date.set(this); date.setAddYearAd(value); date.dateAlone }

fun dateFrom(dateAlone: DateAlone, timeAlone: TimeAlone, existing: Date = Date()): Date {
    return useCalendar {
        it.set(Calendar.YEAR, dateAlone.year)
        it.set(Calendar.MONTH, dateAlone.month - 1)
        it.set(Calendar.DAY_OF_MONTH, dateAlone.day)
        it.set(Calendar.HOUR_OF_DAY, timeAlone.hour)
        it.set(Calendar.MINUTE, timeAlone.minute)
        it.set(Calendar.SECOND, timeAlone.second)
        existing.time = it.timeInMillis
        existing
    }
}

fun Date.format(dateStyle: ClockPartSize, timeStyle: ClockPartSize): String {
    val rawDateStyle = when (dateStyle) {
        ClockPartSize.None -> DateFormat.SHORT
        ClockPartSize.Short -> DateFormat.SHORT
        ClockPartSize.Medium -> DateFormat.MEDIUM
        ClockPartSize.Long -> DateFormat.LONG
        ClockPartSize.Full -> DateFormat.FULL
    }
    val rawTimeStyle = when (timeStyle) {
        ClockPartSize.None -> DateFormat.SHORT
        ClockPartSize.Short -> DateFormat.SHORT
        ClockPartSize.Medium -> DateFormat.MEDIUM
        ClockPartSize.Long -> DateFormat.LONG
        ClockPartSize.Full -> DateFormat.FULL
    }

    val format = if (dateStyle == ClockPartSize.None) {
        if (timeStyle == ClockPartSize.None) {
            throw IllegalStateException()
        }
        DateFormat.getTimeInstance(rawTimeStyle)
    } else if (timeStyle == ClockPartSize.None) {
        DateFormat.getDateInstance(rawDateStyle)
    } else {
        DateFormat.getDateTimeInstance(rawDateStyle, rawTimeStyle)
    }
    return format.format(this)
}

fun Date.iso8601(): String = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").format(this)
fun DateAlone.iso8601(): String = SimpleDateFormat("yyyy-MM-dd").format(dateFrom(this, TimeAlone.noon))
fun TimeAlone.iso8601(): String = SimpleDateFormat("HH:mm:ss").format(dateFrom(Date().dateAlone, this))

object TimeNames {
    private val symbols = DateFormatSymbols()
    val shortMonthNames: List<String> = symbols.shortMonths.toList().dropLast(1)
    val monthNames: List<String> = symbols.months.toList().dropLast(1)
    val shortWeekdayNames: List<String> = symbols.shortWeekdays.toList().drop(1)
    val weekdayNames: List<String> = symbols.weekdays.toList().drop(1)
    fun shortMonthName(oneIndexedPosition: Int): String = shortMonthNames[oneIndexedPosition - 1]
    fun monthName(oneIndexedPosition: Int): String = monthNames[oneIndexedPosition - 1]
    fun shortWeekdayName(oneIndexedPosition: Int): String = shortWeekdayNames[oneIndexedPosition - 1]
    fun weekdayName(oneIndexedPosition: Int): String = weekdayNames[oneIndexedPosition - 1]
}

operator fun TimeAlone.minus(rhs: TimeAlone): TimeAlone {
    val result =
        (this.hour * 60 * 60 + this.minute * 60 + this.second) - (rhs.hour * 60 * 60 + rhs.minute * 60 + rhs.second)
    return if (result < 0) {
        TimeAlone(0, 0, 0)
    } else {
        TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
    }
}

operator fun TimeAlone.plus(rhs: TimeAlone): TimeAlone {
    val result =
        (this.hour * 60 * 60 + this.minute * 60 + this.second) + (rhs.hour * 60 * 60 + rhs.minute * 60 + rhs.second)
    return TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
}

operator fun TimeAlone.minus(rhs: TimeInterval): TimeAlone {
    val result =
        (this.hour * 60 * 60 + this.minute * 60 + this.second) - rhs.milliseconds.toInt() / 1000
    return if (result < 0) {
        TimeAlone(0, 0, 0)
    } else {
        TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
    }
}

operator fun TimeAlone.plus(rhs: TimeInterval): TimeAlone {
    val result =
        (this.hour * 60 * 60 + this.minute * 60 + this.second) + rhs.milliseconds.toInt() / 1000
    return TimeAlone(result / 60 / 60, result / 60 % 60, result % 60)
}
